security/advancedtls/advancedtls.go

```diff
@@ -184,48 +201,96 @@ type ServerOptions struct {
+	// RootCertificateOptions and IdentityCertificateOptions.
+	rootOptionNum := o.RootOptions.nonNilFieldCount()
+	if rootOptionNum > 1 {
+		return nil, fmt.Errorf("at most one field in RootCertificateOptions could be specified")
+	}
+	identityOptionNum := o.IdentityOptions.nonNilFieldCount()
+	if identityOptionNum > 1 {
+		return nil, fmt.Errorf("at most one field in IdentityCertificateOptions could be specified")
+	}
+	if o.IdentityOptions.GetIdentityCertificatesForServer != nil {
+		return nil, fmt.Errorf("GetIdentityCertificatesForServer cannot be specified on the client side")
 	}
-	// We have to set InsecureSkipVerify to true to skip the default checks and
-	// use the verification function we built from buildVerifyFunc.
 	config := &tls.Config{
-		ServerName:           o.ServerNameOverride,
-		Certificates:         o.Certificates,
-		GetClientCertificate: o.GetClientCertificate,
-		InsecureSkipVerify:   true,
+		ServerName: o.ServerNameOverride,
+		// We have to set InsecureSkipVerify to true to skip the default checks and
+		// use the verification function we built from buildVerifyFunc.
+		InsecureSkipVerify: true,
 	}
-	if rootCAs != nil {
-		config.RootCAs = rootCAs
+	// Propagate root-certificate-related fields in tls.Config.
+	switch {
+	case o.RootOptions.RootCACerts != nil:
+		config.RootCAs = o.RootOptions.RootCACerts
+	case o.RootOptions.GetRootCertificates != nil:
+		// In cases when users provide GetRootCertificates callback, since this
+		// callback is not contained in tls.Config, we have nothing to set here.
+		// We will invoke the callback in ClientHandshake.
+	case o.RootOptions.RootProvider != nil:
+		o.RootOptions.GetRootCertificates = func(*GetRootCAsParams) (*GetRootCAsResults, error) {
+			km, err := o.RootOptions.RootProvider.KeyMaterial(context.Background())
+			if err != nil {
+				return nil, err
+			}
+			return &GetRootCAsResults{TrustCerts: km.Roots}, nil
+		}
+	default:
+		// No root certificate options specified by user. Use the certificates
+		// stored in system default path as the last resort.
+		if o.VType != SkipVerification {
+			systemRootCAs, err := x509.SystemCertPool()
+			if err != nil {
+				return nil, err
+			}
+			config.RootCAs = systemRootCAs
+		}
+	}
+	// Propagate identity-certificate-related fields in tls.Config.
+	switch {
+	case o.IdentityOptions.Certificates != nil:
+		config.Certificates = o.IdentityOptions.Certificates
+	case o.IdentityOptions.GetIdentityCertificatesForClient != nil:
+		config.GetClientCertificate = o.IdentityOptions.GetIdentityCertificatesForClient
+	case o.IdentityOptions.IdentityProvider != nil:
+		config.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			km, err := o.IdentityOptions.IdentityProvider.KeyMaterial(context.Background())
+			if err != nil {
+				return nil, err
+			}
+			if len(km.Certs) == 0 {
+				return nil, fmt.Errorf("no identity cert on the client side in IdentityProvider")
+			}
+			if len(km.Certs) > 1 {
```

Comment:
Or alteast you can make a single check instead of two, `len(km.Certs) != 1`, and error out.